src/extensions/scratch3_vizblocks/index.js

```diff
@@ -136,6 +174,346 @@ class Scratch3VizBlocks {
+            const numbers = [];
+
+            for (let l = 0; l <= thisMarker; l++){
+                numbers.push(maxNumber / thisMarker * l);
+            }
+            numbers.forEach((number, index) => {
+                const num = number.toString();
+                for (let i = 0; i < num.length; i++){
+                    this.setLabel(index, i, num[i], axisOption, target);
+                }
+            });
+
+            // set position for input data
+            for (let d = 0; d < thisArray.length; d++) {
+                thisPos.push(thisCenter + (thisArray[d] * this._interval / numbers[1]));
+            }
+
+            // generate internal markers
+            if (axisOption === 'X'){
+                for (let i = 0; i * this._interval <= this._width; i++){
+                    this.runtime.renderer.penLine(penSkinId, penState.penAttributes, this._xCenter + (i * this._interval), this._yCenter, this._xCenter + (i * this._interval), this._yCenter + 5);
+                }
+            } else if (axisOption === 'Y'){
+                for (let j = 1; j * this._interval <= this._height; j++){
+                    this.runtime.renderer.penLine(penSkinId, penState.penAttributes, this._xCenter, this._yCenter + (j * this._interval), this._xCenter + 5, this._yCenter + (j * this._interval));
+                }
+            }
+        }
+    }
+
+    /**
+     * set number text beside the axes
+     * @param {int} numberIndex - the index of label.
+     * @param {int} charIndex - the index of char in label.
+     * @param {int|string} costumeIndex - the index of costume.
+     * @param {string} axisOption - the option of axis, e.g. 'X' or 'Y'.
+     * @param {RenderedTarget} target - target object that has been updated.
+     */
+    setLabel (numberIndex, charIndex, costumeIndex, axisOption, target){
+        const penSkinId = this._getPenLayerID();
+        let xPos;
+        let yPos;
+        let size;
+        let direction;
+
+        // set text based on costumeIndex
+        if (costumeIndex <= 9 && costumeIndex >= 0){
+            if (axisOption === 'X'){
+                xPos = this._xCenter + (numberIndex * this._interval) + (charIndex * 8);
+                yPos = this._yCenter - 10;
+            } else if (axisOption === 'Y') {
+                xPos = this._xCenter - 10;
+                yPos = this._yCenter + (numberIndex * this._interval) + (charIndex * 8);
+            }
+            size = 20;
+        } else if (costumeIndex >= 'A' && costumeIndex <= 'Z') {
+            if (axisOption === 'X'){
+                direction = 90;
+                xPos = this._xCenter + (this._interval * 2) + (numberIndex * this._interval / 2);
+                yPos = this._yCenter - 40;
+            } else if (axisOption === 'Y') {
+                direction = 0;
+                xPos = this._xCenter - 40;
+                yPos = this._yCenter + (this._interval * 2) + (numberIndex * this._interval / 2);
```

Comment:
Let's start from the baseline to accommodate more room for long labels.

```suggestion
                yPos = this._yCenter + (numberIndex * this._interval / 2);
```